src/js/components/award/shared/activity/ContractGrantsActivityChart.jsx

```diff
@@ -97,23 +121,117 @@ const ContractGrantsActivityChart = ({
-    //         .range([0, visualizationWidth])
-    //         .nice();
-    //     setXTicks(updatedTicksWithSpacing);
-    //     setXScale(() => updatedScale);
-    // }, [xDomain, visualizationWidth]);
+    const createXScaleAndTicks = useCallback(() => {
+        if (!xDomain.length) return null;
+        /**
+         * By design, we want to leave space between the start of the chart
+         * and the end of the chart.
+         */
+        const spacing = visualizationWidth * 0.05;
+        const scale = scaleLinear()
+            .domain(xDomain)
+            .range([spacing, visualizationWidth - spacing - padding.left]);
+        let theTicks = scale.ticks(6);
+        // add first and last tic per design
+        theTicks.splice(0, 0, xDomain[0]); // first tick
+        theTicks.push(xDomain[1]); // last tick
+
+        const differences = theTicks.reverse().reduce((acc, tick, i, src) => {
+            const nextTick = src[i + 1];
+            if (!nextTick) return acc; // last tick ignore
+            acc.push(tick - nextTick);
+            return acc;
+        }, []);
+
+        const averageDifference = sum(differences) / theTicks.length;
+        // ascending order since we reversed above
+        theTicks.reverse();
+        /**
+         * manuallyCreateXTicks
+         * - Manually create the x ticks starting with the first tick and
+         * adding the difference sequentially.
+         */
+        const manuallyCreateXTicks = (ticks, diff) => {
+            const newTicks = cloneDeep(ticks);
+            /**
+             * Tthe ticks method for d3 does not include ticks for the very beginning
+             * and the end of the domain. This leaves a gap between the last tick we create
+             * and the last tick of the domain. Therefore we increase our array by two.
+             * Since we ignore the first and last tick we cannot unshift or push :(.
+             */
+            newTicks.splice(1, 0, 1);
+            return newTicks.reduce((acc, tick, i) => {
+                // ignore first and last tick
+                if (i === 0 || i + 1 === newTicks.length) {
+                    acc.push(tick);
+                }
+                else {
+                    acc.push(acc[i - 1] + diff);
+                }
+                return acc;
+            }, []);
+        };
+        /**
+         * removeOverlappingTicks
+         * - Manually adding ticks could cause overlap since d3 is creating their own ticks.
+         * This filters any ticks that might overlap.  The number 20 was derived
+         * from visual evidence of overlap on this award ASST_NON_1905OH5MAP_7530.
+         * @param {Number[]} - array of milliseconds
+         * @param {function} - d3 scale function
+         * @returns {Number[]} - array of milliseconds
+         */
+        let skipNext = false;
+        const removeOverlappingTicks = (ticks, scaleFunc) => ticks.reduce((acc, tick, i, src) => {
```

Comment:
Reccomend using .filter for this purpose. S/O to `.reduce` nonetheless. #thagoat

Reply:
I believe I started with filter, lol, as I like using all the array methods. I think I needed to use reduce due to the logic in the function.

Reply:
@jonhill13 if all you're doing is removing overlapping ticks, like the name of the function suggests, all you're doing is filtering the array.

```suggest
const removeOverlappingTicks = (ticks, scaleFunc) => ticks.filter((tick, i, src) => {
           if (skipNext) {
                skipNext = false;
                return false;
            }
            const currentTick = scaleFunc(tick);
            const nextTick = scaleFunc(src[i + 1]);
            if (nextTick) {
                const difference = nextTick - currentTick;
                if (difference <= 20) { // ticks overlap
                    // keeps last tick
                    if (i + 2 === src.length) {
                        return false;
                    }
                    skipNext = true;
                }
                return true;
            }
            return true; // first tick
}
```

Reply:
Yeah

Reply:
done https://github.com/fedspendingtransparency/usaspending-website/pull/1559